### PR TITLE
Implement passport strategies

### PR DESCRIPTION
**Pull request by deepakrkris**
* create adapter for passport strategies

* create unit test for adapter

https://github.com/strongloop/loopback-next/issues/138

cc @raymondfeng

### CODE REVIEW

File: packages/authentication/test/unit/strategy-adapter.test.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  describe('authenticate()', () => {
+
+    it('calls the authenticate method of the strategy', (done) => {
+      const strategy = new passportStrategy();
+      let calledFlag = false;
+      strategy.authenticate = function() {
+        calledFlag = true;
+      };
```
Comment (bajtos) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: Lines 24-27 can be rewritten using a sinon spy. Not a big deal, just FYI.

File: packages/authentication/test/unit/strategy-adapter.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+describe('Strategy Adapter', () => {
+  const mockUser: User = {id: 'mock-user', role: 'mock-role'};
+  interface User {
+    id: string;
+    role: string;
+  }
+
+  describe('authenticate()', () => {
+
+    it('calls the authenticate method of the strategy', (done) => {
```
Comment (bajtos) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: AFAICT, this test runs synchronously, there is no need to use `function (done)` here.

File: packages/authentication/test/unit/strategy-adapter.test.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      done();
+    });
+
+    it('adds success handler to the strategy', (done) => {
+      const strategy = new MockStrategy();
+      const adapter = new StrategyAdapter(strategy);
+      const request = <ParsedRequest> {};
+      // tslint:disable-next-line:no-floating-promises
+      adapter.authenticate(request)
+      .then((user: User) => {
```
Comment (bajtos) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: This does not work correctly when `.authenticate` returns a rejected promise - the rejected promise will end up in the background and eventually got reported via Node.js unhandled promise reporter. This is one of the reasons why `tslint` complained about a floating promise. Hiding the tslint error is not a proper fix ;-)

In general, don't mix `done` callback-style and promise-style code in tests. Mocha supports promises OOTB. We should use async/await to get even cleaner code.

```ts
it('adds success handler to the strategy', async () => {
  // ...
  const user = await adapter.authenticate(request);
  expect(user)//...
});
```

File: packages/authentication/src/strategy-adapter.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ * Interface definition of a passport strategy.
+ */
+export interface Strategy {
+  authenticate(req: http.ServerRequest): void;
+}
+
+/**
+ * Shimmed Request to satisfy express requirements of passport strategies.
+ */
+export class ShimRequest {
```
Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: This is a place holder for any shim required as we get to test more passport strategies through the community

Comment (bajtos) [inline packages/authentication/src/strategy-adapter.ts]: Fair enough. Let's keep the current design then, we can easily improve it later if there is a need for that.

File: packages/authentication/src/strategy-adapter.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      const self = this;
+
+      // add success state handler to strategy instance
+      strategy.success = function(user: object) {
+        resolve(user);
+      };
+
+      // add failure state handler to strategy instance
+      strategy.fail = function(challenge: string) {
+        reject(new Error(challenge));
```
Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: adding http-errors to any other package gives an eslint error ``` Error at packages/authentication/node_modules/@types/http-errors/index.d.ts:96:5: Duplicate identifier 'export='.
Error at packages/example-codehub/node_modules/@types/http-errors/index.d.ts:96:5: Duplicate identifier 'export='. ```

Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: I have an interim fix as per your suggestion 
```
in the worst case, I think we can reference http-errors from `@loopback/core` only and export it for use in authentication and example-codehub (or any user app)
```

File: packages/authentication/test/unit/strategy-adapter.test.ts (hunk 2 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: it does not run synchronously. authenticate in mock strategy calls process.nexttick
```
authenticate(req: ParsedRequest) {
      process.nextTick(this.returnMockUser.bind(this));
    }
```

Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: I see your point, I was not using await to call the async function, ok fixed now

File: packages/authentication/test/unit/strategy-adapter.test.ts (hunk 1 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: I will open a separate issue for this

File: packages/authentication/test/unit/strategy-adapter.test.ts (hunk 3 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: agreed, fixed now

**Comment by deepakrkris** [comment]
cc @bajtos @raymondfeng

File: packages/authentication/src/index.ts [hunk 6]
```diff
@@ -2,5 +2,7 @@
 // Node module: @loopback/authentication
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
+// NOTE(bajtos) This file is used by VSCode/TypeScriptServer at dev time only
```
Comment (bajtos) [inline packages/authentication/src/index.ts]: This comment applies to `packages/authentication/index.ts` only, please revert it here.

File: packages/authentication/src/strategy-adapter.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ * Shimmed Request to satisfy express requirements of passport strategies.
+ */
+export class ShimRequest {
+  headers: Object;
+  query: Object;
+  url: string;
+  path: string;
+  method: string;
+  constructor(request?: ParsedRequest) {
+    if (request) {
```
Comment (bajtos) [inline packages/authentication/src/strategy-adapter.ts]: Nitpick: use early return to reduce level of indentation.

```ts
constructor(request?: ParsedRequest) {
  if (!request) return;
  this.headers = request.headers;
  // ...
}
```

File: packages/authentication/test/unit/strategy-adapter.test.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    it('calls the authenticate method of the strategy', () => {
+      const strategy = new passportStrategy();
+      let calledFlag = false;
+      // TODO: (as suggested by @bajtos) use sinon spy
+      strategy.authenticate = function() {
+        calledFlag = true;
+      };
+      const adapter = new StrategyAdapter(strategy);
+      const request = <ParsedRequest> {};
+      // tslint:disable-next-line:no-floating-promises
```
Comment (bajtos) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: This tslint exception should not be needed anymore, can you remove it please?

File: packages/authentication/test/unit/strategy-adapter.test.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      const user: Object = adapter.authenticate(request);
+      expect(calledFlag).to.be.true();
+    });
+
+    it('returns a promise which resolves to an object', async () => {
+      const strategy = new MockStrategy();
+      const adapter = new StrategyAdapter(strategy);
+      const request = <ParsedRequest> {};
+      // tslint:disable-next-line:no-floating-promises
```
Comment (bajtos) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: This tslint exception should not be needed anymore, can you remove it please?

File: packages/core/src/index.ts [hunk 10]
```diff
@@ -20,3 +20,4 @@ export {ServerRequest, ServerResponse} from 'http';
 // internals used by unit-tests
 export {parseOperationArgs} from './parser';
 export {ParsedRequest, parseRequestUrl} from './router/SwaggerRouter';
+export {NotFound as HttpErrorNotFound} from 'http-errors';
```
Comment (bajtos) [inline packages/core/src/index.ts]: IMO, exporting only `NotFound` is too limiting. I am proposing to export full `HttpError` constructor/object.

```ts
// here
export * as HttpError from 'http-errors';

// usage
import {HttpError} from '@loopback/core';
throw new HttpError.NotFound('Current user not found (?!).');
```

File: packages/authentication/src/strategy-adapter.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+      const self = this;
+
+      // add success state handler to strategy instance
+      strategy.success = function(user: object) {
+        resolve(user);
+      };
+
+      // add failure state handler to strategy instance
+      strategy.fail = function(challenge: string) {
+        reject(new HttpErrorNotFound(challenge));
```
Comment (bajtos) [inline packages/authentication/src/strategy-adapter.ts]: This is returning 404 Not Found, are we sure that's the correct status code? My expectation is that the server should send 401 Not Authorized with WWW-Authenticate header field containing a challenge applicable to the requested resource.

@raymondfeng do you have any opinion on this?

File: packages/authentication/test/unit/strategy-adapter.test.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+      const strategy = new passportStrategy();
+      let calledFlag = false;
+      // TODO: (as suggested by @bajtos) use sinon spy
+      strategy.authenticate = function() {
+        calledFlag = true;
+      };
+      const adapter = new StrategyAdapter(strategy);
+      const request = <ParsedRequest> {};
+      // tslint:disable-next-line:no-floating-promises
+      const user: Object = adapter.authenticate(request);
```
Comment (bajtos) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: > it does not run synchronously. authenticate in mock strategy calls process.nexttick

I think you need to `await` the result of `adapter.authenticate(request)` then, similarly to what you are doing in the next test.

File: packages/authentication/src/strategy-adapter.ts (hunk 11 above)
Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: agreed it should be 404 unauthorized

File: packages/authentication/test/unit/strategy-adapter.test.ts (hunk 8 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: agreed

File: packages/authentication/test/unit/strategy-adapter.test.ts (hunk 9 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: agreed

File: packages/core/src/index.ts (hunk 10 above)
Comment (deepakrkris) [inline packages/core/src/index.ts]: agreed, but the syntax is as follows
```
// import all errors from external http-errors package
import * as HttpErrors from 'http-errors';

// http errors
export {HttpErrors};
```

File: packages/authentication/test/unit/strategy-adapter.test.ts (hunk 12 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/strategy-adapter.test.ts]: agreed

File: packages/authentication/src/index.ts (hunk 6 above)
Comment (deepakrkris) [inline packages/authentication/src/index.ts]: agreed

File: packages/authentication/src/strategy-adapter.ts (hunk 7 above)
Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: agreed

File: packages/authentication/src/strategy-adapter.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+// License text available at https://opensource.org/licenses/MIT
+import * as http from 'http';
+import {ServerRequest as Request} from 'http';
+import {HttpErrors, ParsedRequest} from '@loopback/core';
+
+/**
+ * Interface definition of a passport strategy.
+ */
+export interface Strategy {
+  authenticate(req: http.ServerRequest): void;
```
Comment (raymondfeng) [inline packages/authentication/src/strategy-adapter.ts]: Don't we expect a Promise of the result? Typically, authentication will either fail or establish a user.

Comment (raymondfeng) [inline packages/authentication/src/strategy-adapter.ts]: The passport `Strategy.prototype.authenticate` also takes an `options` object.

File: packages/authentication/src/strategy-adapter.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+ *   2. provides shimming of requests for passport authentication
+ *   3. provides lifecycle similar to express to the passport-strategy
+ *   3. provides state methods to the strategy instance
+ * see: https://github.com/jaredhanson/passport
+ */
+export class StrategyAdapter {
+  /**
+   * @param strategyCtor function which implements a passport-strategy;see: http://passportjs.org/
+   */
+  constructor(private strategyCtor: Strategy) {}
```
Comment (raymondfeng) [inline packages/authentication/src/strategy-adapter.ts]: The parameter name should be `strategy` as its an instance of `Strategy`, not the class/constructor.

File: packages/authentication/test/unit/fixtures/mock-strategy.ts [hunk 15]
```diff
@@ -0,0 +1,59 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import {ParsedRequest} from '@loopback/core';
+
+/**
+ * Test fixture for a mock asynchronous passport-strategy
+ */
+export class MockStrategy {
```
Comment (raymondfeng) [inline packages/authentication/test/unit/fixtures/mock-strategy.ts]: `implements Strategy`?

File: packages/authentication/test/unit/fixtures/mock-strategy.ts [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+  private mockUser: Object;
+  constructor() {}
+  setMockUser(userObj: Object) {
+    this.mockUser = userObj;
+  }
+  /**
+   * authenticate() function similar to passport-strategy packages
+   * @param req
+   */
+  async authenticate(req: ParsedRequest) {
```
Comment (raymondfeng) [inline packages/authentication/test/unit/fixtures/mock-strategy.ts]: This signature is different from the `Strategy.authenticate`.

File: packages/authentication/src/strategy-adapter.ts (hunk 13 above)
Comment (raymondfeng) [inline packages/authentication/src/strategy-adapter.ts]: Can you import [@types/passport](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/passport) directly?

Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: agreed

Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: that's a nice idea

File: packages/authentication/src/strategy-adapter.ts (hunk 14 above)
Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: the passed in Strategy function is used as constructor to do a new Strategy(), so strategyCtor is more appropriate

Comment (deepakrkris) [inline packages/authentication/src/strategy-adapter.ts]: I think strategyClass would be an appropriate name, let me change that

File: packages/authentication/test/unit/fixtures/mock-strategy.ts (hunk 15 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/fixtures/mock-strategy.ts]: agreed

File: packages/authentication/test/unit/fixtures/mock-strategy.ts (hunk 16 above)
Comment (deepakrkris) [inline packages/authentication/test/unit/fixtures/mock-strategy.ts]: agreed

**Comment by deepakrkris** [comment]
thanks guys

**Pull request merged**